draft-bbs-signatures.md

```diff
@@ -680,38 +680,56 @@ Procedure:
 
 # Utility Operations
 
-## CreateGenerators
+## Generator point computation
 
-The CreateGenerators operation defines how to create a set of generators that form a part of the public parameters used by the BBS Signature scheme to accomplish operations such as Sign, Verify, ProofGen and ProofVerify.
+The `create_generators` operation defines how to create a set of generators that form a part of the public parameters used by the BBS Signature scheme to accomplish operations such as Sign, Verify, ProofGen and ProofVerify. It takes one input, the number of generator points to create, which is determined in part by the number of selectively-revealed signed messages.
+
+As an optimization, implementations MAY cache the result of `create_generators` for a specific `generator_seed` (determined by the ciphersuite) and `count`. The values `c` and `v` MAY also be cached in order to efficiently extend a existing list of generator points.
 
 ```
-generators = CreateGenerators(dst, message_generator_seed, length);
+generators = create_generators(count)
 
 Inputs:
 
-- dst (REQUIRED), octet string. Domain Separation Tag.
-- message_generator_seed (REQUIRED), octet string.
-- length (REQUIRED), unsigned integer. Number of generators to create from the seed and dst.
+- count (REQUIRED), unsigned integer. Number of generator points to create.
+
+Parameters:
+
+- generator_seed, octet string. A seed value selected by the ciphersuite.
+- hash_to_curve_g1. A hash-to-curve suite selected by the ciphersuite.
+
+Definitions:
+
+- expand_message is the corresponding operation associated with the ciphersuite's hash_to_curve_g1 parameter.
+- seed_dst is the octet string representing "BBS-SIG-GENERATOR-SEED" in the ASCII character encoding.
+- hash_output_bytes is the integer given by ceil((log2(r) + k) / 8).
+- generator_dst is the octet string representing "BBS-SIG-GENERATOR-DST" in the ASCII character encoding.
 
 Outputs:
 
 - generators, an array of generators.
 
 Procedure:
 
-1. h = xof(seed)
+1. n = 1
+
+2. v = expand_message(generator_seed, seed_dst, hash_output_bytes)
+
+3. for i in range(1, count):
 
-2. for i in 0 to length:
+4.    generator_i = Identity_G1
 
-3.    generator_i = Identity_G1
+5.    while generator_i == Identity_G1 or generator_i == P1:
 
-4.    while(generator_i == Identity_G1 or generator_i == P1)
+6.        candidate = hash_to_curve_g1(v, generator_dst)
 
-5.        candidate = hash_to_curve_g1(h.read(xof_no_of_bytes), dst)
+7.        if candidate not in generators[1..i-1]: generator_i = candidate
 
-6.        if candidate not in generators: generator_i = candidate
+8.        v = expand_message(v || I2OSP(n, 4), seed_dst, hash_output_bytes)
```

Comment:
> generator_2 is not indistinguishable from random if you know generator_1

Well, this assumes that generator_1 is uniquely determined from the generator_seed, but `v` could be made large enough to ensure there are collisions in the hash-to-curve output. The generator points are not used as inputs to create the next state.

Another option would be to use `v[X..]` as the hash-to-curve input, so the first X bytes remain as 'private' for the generator state machine.

Reply:
Here's what the last option might look like:

```
private_bytes = 16  # could be in definitions, probably does not need to be curve dependent
hash_output_bytes = private_bytes + expand_length
n = 1
v = expand_message(generator_seed, seed_dst, hash_output_bytes)
for i in range(1, count):
   generator_i = Identity_G1
   while generator_i == Identity_G1 or generator_i == P1:
       candidate = hash_to_curve_g1(v[private_bytes..hash_output_bytes], generator_dst)
       if candidate not in generators[1..i-1]: generator_i = candidate
       v = expand_message(v || I2OSP(n, 4), seed_dst, hash_output_bytes)
       n = n + 1
return generators
```

Another option is to initialize v[..private_bytes] with zeroes when hashing the seed.

Reply:
> generator_i does not determine the generator_(i+1) though, as the generator points are not used as inputs to create the next state. There are multiple v values which could produce each generator point, especially if we increase the size of v. And of course you would have to invert the hash somehow to predict the next value.

Yea this is true in practice, however it may not be "strong" enough for a security proof. Proving "statistical closeness to uniform" is hard, since`P(generator_i | generator_(i-1), ..., generator_1) >> 1/i`. Also, I think we should treat the seed as a known, fixed value (like an IV) in this case, which does not help with the proof. Even if that was not the case though, it is true that multiple `v`'s can give the same point, but a simulator cannot find them, if we assume coalition resistance. As you noted though, we may be able to prove this for "computationally close to uniform".

Personally, I like the cut-off method. A prefix-free method could also work and is pretty close to the current direction of the PR. That said, I do suggest leaving any updates (if there are any required) after we do a formal analysis. I will try to write down a formal proof, and if there are any updates required as a result, we can make them in following PRs.